CONSCENDI/generate_buses.py

```diff
@@ -0,0 +1,340 @@
+import pandas as pd
+import numpy as np
+from typing import Tuple
+from constants import create_logger, load_template, query_and_retry, cycle_api_key, KEYS
+import random
+import time
+from itertools import cycle
+from datetime import datetime
+import wandb
+import ast
+
+now = datetime.now()
+display_name = f"generate_violations{now.month}-{now.day}_{now.hour}:{now.minute}"
+run = wandb.init(project='context-distillation', entity='curai', name=display_name)
+
+def _generate_violation_scenarios(rule, model, logger, scenario_path, domain, n=5):
+    """
+    Generate a list of n likely scenarios that may induce the patient to activate the guardrails
+
+    Ex: 
+    logger = create_logger()
+    _generate_violation_scenarios("Curai is not allowed to serve when patients request assistance for other people. ", "text-alpha-002-current", logger, n=10)
+
+    This would generate 10 likely scenarios that would break the rule of serving when patients request assistance for other people.
+    """
+    # prompt = "You are a virtual assistant at an adult telehealth medical clinic called Curai in the USA. This is a rule you cannot break. \n[Rule: "
+    # prompt += rule + "]\n"
+    # prompt += f"What are {n} certain ways that Curai would find out in text conversations that a patient is attemping to break this rule? Separate each with a newline \n-"
+    # print(prompt)
+    prompt = load_template(scenario_path).render(
+                rule = rule,
+                n=n,
+                domain=domain
+                )
+    scenarios = query_and_retry(
+        formatted_prompt = prompt,
+        engine = model,
+        temperature = 0.25, 
+        max_tokens = 75 * n,
+        stop=["[STOP]", "STOP"],
+        logger=logger
+    )['choices'][0]['message']['content']
+
+    scenarios = scenarios.rstrip('\n [STOP]')
+    scenarios_dict = ast.literal_eval(scenarios)
+
+    return scenarios_dict
+
+def generate_violations(
+    violation_path: str, # jinja file
+    rules_path: str, # csv file
+    output_path: str, 
+    model: str, 
+    num_samples: int = 300,
+    num_scenarios: int = 10,
+    scenario_csv_path: str = None,
+    scenario_prompt_path: str = None, # jinja file
+    domain: str = None,
+    # scenario_path: str = None
+) -> Tuple[pd.DataFrame, str]:
+    """
+    purpose: Generate test samples given a particular model.
+    
+    Example: generate(
+            "multi-rule-guardrail/prompts/oob_violation.jinja", 
+            "multi-rule-guardrail/prompts/oob_nonviolation.jinja", 
+            "multi-rule-guardrail/prompts/oob_rules.csv", 
+            "text-alpha-002-current", 
+            10,
+            5)
+    
+    rules: csv file with one column. title should be rule and there should be a new line between each rule. 
+
+    ===Rules Example:==== 
```

Comment:
rremove all references to Curai. We can remove the entirety of Curai example here